File: app/modules/projects/schemas.py
from __future__ import annotations

from datetime import datetime
from typing import Literal

from app.modules.shared.schemas import DashboardModel


class ProjectEntry(DashboardModel):
    id: str
    name: str
    description: str | None
    project_url: str | None
    github_repo_url: str | None
    project_path: str | None
    sandbox_mode: Literal["read-only", "workspace-write", "danger-full-access"]
    git_branch: str | None
    created_at: datetime
    updated_at: datetime


class ProjectsResponse(DashboardModel):
    entries: list[ProjectEntry]


class ProjectPlanLinkEntry(DashboardModel):
    project_id: str
    plan_count: int
    latest_plan_slug: str | None
    latest_plan_updated_at: datetime | None


class ProjectPlanLinksResponse(DashboardModel):
    entries: list[ProjectPlanLinkEntry]


class ProjectCreateRequest(DashboardModel):
    name: str
    description: str | None = None
    project_url: str | None = None
    github_repo_url: str | None = None
    project_path: str | None = None
    sandbox_mode: str | None = None
    git_branch: str | None = None


class ProjectUpdateRequest(DashboardModel):
    name: str
    description: str | None = None
    project_url: str | None = None
    github_repo_url: str | None = None
    project_path: str | None = None
    sandbox_mode: str | None = None
    git_branch: str | None = None


class ProjectDeleteResponse(DashboardModel):
    status: str


class ProjectOpenFolderRequest(DashboardModel):
    target: Literal["vscode", "file-manager"] = "vscode"


class ProjectOpenFolderResponse(DashboardModel):
    status: str
    project_path: str
    target: Literal["vscode", "file-manager"] = "vscode"
    editor: str | None = None

File: apps/frontend/src/features/projects/schemas.ts
import { z } from "zod";

export const PROJECT_SANDBOX_MODES = [
  "read-only",
  "workspace-write",
  "danger-full-access",
] as const;

export const ProjectSandboxModeSchema = z.enum(PROJECT_SANDBOX_MODES);

export const ProjectEntrySchema = z.object({
  id: z.string().min(1),
  name: z.string().min(1),
  description: z.string().nullable(),
  projectUrl: z.string().nullable(),
<<<<<<< Updated upstream
  githubRepoUrl: z.string().nullable().default(null),
=======
  githubRepoUrl: z.string().nullable(),
>>>>>>> Stashed changes
  projectPath: z.string().nullable(),
  sandboxMode: ProjectSandboxModeSchema,
  gitBranch: z.string().nullable(),
  createdAt: z.string().datetime({ offset: true }),
  updatedAt: z.string().datetime({ offset: true }),
});

export const ProjectsResponseSchema = z.object({
  entries: z.array(ProjectEntrySchema).default([]),
});

export const ProjectCreateRequestSchema = z.object({
  name: z.string().min(1).max(128),
  description: z.string().max(512).nullable().optional(),
  projectUrl: z.string().max(2048).nullable().optional(),
  githubRepoUrl: z.string().max(2048).nullable().optional(),
  projectPath: z.string().max(1024).nullable().optional(),
  sandboxMode: ProjectSandboxModeSchema.default("workspace-write"),
  gitBranch: z.string().max(255).nullable().optional(),
});

export const ProjectUpdateRequestSchema = z.object({
  name: z.string().min(1).max(128),
  description: z.string().max(512).nullable().optional(),
  projectUrl: z.string().max(2048).nullable().optional(),
  githubRepoUrl: z.string().max(2048).nullable().optional(),
  projectPath: z.string().max(1024).nullable().optional(),
  sandboxMode: ProjectSandboxModeSchema.default("workspace-write"),
  gitBranch: z.string().max(255).nullable().optional(),
});

export const ProjectDeleteResponseSchema = z.object({
  status: z.string().min(1),
});

export const ProjectOpenFolderResponseSchema = z.object({
  status: z.string().min(1),
  projectPath: z.string().min(1),
  editor: z.string().nullable().optional(),
});

export const ProjectPlanLinkEntrySchema = z.object({
  projectId: z.string().min(1),
  planCount: z.number().int().min(0),
  latestPlanSlug: z.string().nullable(),
  latestPlanUpdatedAt: z.string().datetime({ offset: true }).nullable(),
});

export const ProjectPlanLinksResponseSchema = z.object({
  entries: z.array(ProjectPlanLinkEntrySchema).default([]),
});

export type ProjectEntry = z.infer<typeof ProjectEntrySchema>;
export type ProjectsResponse = z.infer<typeof ProjectsResponseSchema>;
export type ProjectCreateRequest = z.input<typeof ProjectCreateRequestSchema>;
export type ProjectUpdateRequest = z.input<typeof ProjectUpdateRequestSchema>;
export type ProjectDeleteResponse = z.infer<typeof ProjectDeleteResponseSchema>;
export type ProjectOpenFolderResponse = z.infer<typeof ProjectOpenFolderResponseSchema>;
export type ProjectSandboxMode = z.infer<typeof ProjectSandboxModeSchema>;
export type ProjectPlanLinkEntry = z.infer<typeof ProjectPlanLinkEntrySchema>;
export type ProjectPlanLinksResponse = z.infer<typeof ProjectPlanLinksResponseSchema>;
